onnxruntime/core/optimizer/propagate_cast_ops.cc

```diff
@@ -0,0 +1,709 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+#include "core/optimizer/initializer.h"
+#include "core/optimizer/propagate_cast_ops.h"
+#include "core/graph/graph_utils.h"
+#include "core/optimizer/utils.h"
+#include <deque>
+
+using namespace ONNX_NAMESPACE;
+using namespace onnxruntime::common;
+namespace onnxruntime {
+// The collection fp16_allow_ops, specifies for a given propagate_cast_ops level, a vector of node op_types that
+// the code is allowed to propage Cast operations cross. The user may specify a custom list of optypes using level 0.
+// The opcodes are split into multiple levels. Cast propagation is done based on the level. Level 2 op code
+// list includes Level 1 list also.
+static std::vector<std::unordered_set<std::string>> fp16_allow_ops = {
+    /* Level 0 */ {},
+    /* Level 1 */ {"Transpose", "Reshape", "Split", "Relu", "BiasFastGelu", "BiasGelu", "LayerNorm", "Gelu", "FastGelu", "Tanh"},
```

Comment:
>Tanh [](start = 122, length = 4)

Tanh fp16 is indeed performed in float, see 
onnxruntime/contrib_ops/cuda/activation/activations_impl.cu
onnxruntime/core/providers/rocm/cu_inc/common.cuh

template <>
__device__ __inline__ half _Tanh(half a) { return half(tanhf((float)a)); }

Reply:
but Gelu fp16, is not completely identical to CastTo16(GeluIn32(CastTo32(x)))

see onnxruntime/core/providers/rocm/cu_inc/common.cuh

template <>
__device__ __inline__ half _Normcdf(half a) { return half(normcdff((float)a)); }

template <typename T>
__device__ __inline__ T _Gelu(T a) {
  return a * _Normcdf(a);
}

---
In reply to: [612001860](https://github.com/microsoft/onnxruntime/pull/7191#discussion_r612001860) [](ancestors = 612001860)

Reply:
Same for BiasGelu, and FastGelu. 
So, strictly speaking, they should all be in Level 2. 

I think we should follow our guide line on level1 and level2: anything we put in Level1 has to make no changes on the numerical results. 

For the model owner to invoke the Gelu, he/she has to use set level = 1 and "Gelu in allow-list"

---
In reply to: [612003429](https://github.com/microsoft/onnxruntime/pull/7191#discussion_r612003429) [](ancestors = 612003429,612001860)